fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/JoinLeftAsscomRule.java

```diff
@@ -0,0 +1,192 @@
+        List<ColumnRefSet> splitCols = JoinReorderHelper.splitJoinOutputCols(bottomJoinExpr);
+        topJoinLeftInputColSet.union(splitCols.get(0));
+        topJoinLeftInputColSet.union(colSetC);
+
+
+        JoinReorderHelper.splitCondition(topJoin.getOnPredicate(),
+                topJoinLeftInputColSet, bottom, top);
+        JoinReorderHelper.splitCondition(bottomJoin.getOnPredicate(),
+                newBottomInputColSet, bottom, top);
+
+        ScalarOperator newTopOnCondition = Utils.compoundAnd(top);
+        ScalarOperator newBottomOnCondition = Utils.compoundAnd(bottom);
+
+        if (MapUtils.isNotEmpty(projects.get(1)) && !(optExprB.getOp() instanceof LogicalScanOperator)
+                && optExprB.getInputs().isEmpty()) {
+            // when optExprB is not scan and inputs is empty, this merging projection will yield a new OptExpression.
+            // Hence, an error optExpression without children will be inserted into the memo.
+            // TODO fix it in the future
+            return Collections.emptyList();
+        }
+
+        // merge two projections of optExprB
+        if (MapUtils.isNotEmpty(projects.get(1))) {
+            Map<ColumnRefOperator, ScalarOperator> inputProject;
+
+            if (optExprB.getOp().getProjection() != null) {
+                inputProject = optExprB.getOp().getProjection().getColumnRefMap();
+
+            } else {
+                inputProject = colSetB.getColumnRefOperators(columnRefFactory).stream()
+                        .collect(Collectors.toMap(Function.identity(), Function.identity()));
+            }
+
+            Map<ColumnRefOperator, ScalarOperator> newProject = JoinReorderHelper.mergeProject(projects.get(1),
+                    inputProject);
+
+            Operator.Builder builder = OperatorBuilderFactory.build(optExprB.getOp());
+
+            Operator newRightChildOp = builder.withOperator(optExprB.getOp())
+                    .setProjection(new Projection(newProject)).build();
+            optExprB = OptExpression.create(newRightChildOp, optExprB.getInputs());
+        }
+
+        LogicalJoinOperator newTopJoin = newTopJoinBuilder.withOperator(topJoin)
+                .setJoinType(bottomJoin.getJoinType())
+                .setOnPredicate(newTopOnCondition)
+                .setPredicate(newTopPredicate)
+                .build();
+
+        // prune unnecessary bottom output columns according to the projection of newTopJoin
+        if (newTopJoin.getProjection() != null) {
+            newBottomInputColSet.and(newTopJoin.getRequiredChildInputColumns());
+        }
+
+        Map<ColumnRefOperator, ScalarOperator> newBottomProjection = JoinReorderHelper.mergeProject(projects.get(0),
+                newBottomInputColSet.getColumnRefOperators(columnRefFactory).stream()
+                        .collect(Collectors.toMap(Function.identity(), Function.identity())));
+        Projection newBottomJoinProject = MapUtils.isEmpty(newBottomProjection) ?
```

Comment:
What's the difference of null and an empty `Projection` ? Shall we unify it into empty `Projection` ?

Reply:
Many operator used `getOutputColumns` as below. If you set a empty `Projection`, you may get a wrong output columns. I agree with your point, we need refactor it and unify empty Projection and `null` Projection.
```
public ColumnRefSet getOutputColumns(ExpressionContext expressionContext) {
        if (projection != null) {
            return new ColumnRefSet(projection.getOutputColumns());
        } else {
            ColumnRefSet columns = new ColumnRefSet();
            for (int i = 0; i < expressionContext.arity(); ++i) {
                columns.union(expressionContext.getChildLogicalProperty(i).getOutputColumns());
            }
            return columns;
        }
    }
```